content/content/oop.md

```diff
@@ -4,9 +4,108 @@ title: Object Oriented Programming
 
 # Object Oriented Programming
 
-Object oriented programming is supported through inheritance and methods, which are virtual procedures. Methods work in the same way as [procs](/procs/), including UFCS, but the runtime type of the object that a method is called with is used to determine which version to call.
+## Member functions or Method calls
 
-Inheritance is created with the `of` keyword in the type declaration of an object. A method is overridden by creating a new method with parameter types of the subtype. 
+Member functions, typically known as methods in OOP terminology, can be implemented directly using `proc`s, thanks to the Uniform Call Syntax.
+
+``` nimrod
+type Animal = object
+  name: string
+  age: int
+  
+proc speak(self: Animal, msg: string) = 
+  echo self.name & " says:" & msg 
+  
+var sparky = Animal(name: "Sparky", age: 10)
+
+sparky.speak("Hi")
+# this is the same as the below
+speak(sparky, "Hi")
+```
+
+As you might have guessed, the method call syntax is not restricted to object types.
+
+``` nimrod
+proc double(num: int):int =
+  return num*2
+  
+double(10)
+# same as
+10.double()
+# you can even drop the empty paranthesis
+10.double
+```
+
+Now, when using UFCS to think about `proc`s as member functions, there is an important point to keep in mind about mutability. Consider this example.
+
+``` nimrod
+type Animal = object
+  name: string
+  age: int
+  
+proc incAge(self: Animal) =
+  self.age += 1
+  
+var sparky = Animal(name: "Sparky", age: 3)
+```
+
+This fails compiling with the following error:
+``` console
+Error: type mismatch: got <int, int literal(1)>
+but expected one of:
+proc `+=`[T: SomeInteger](x: var T; y: T)
+  first type mismatch at position: 1
+  required type for x: var T: SomeInteger
+  but expression 'self.age' is immutable, not 'var'
+proc `+=`[T: float | float32 | float64](x: var T; y: T)
+  first type mismatch at position: 1
+  required type for x: var T: float or float32 or float64
+  but expression 'self.age' is of type: int
+
+expression: self.age += 1
+```
+
+If you try something simpler:
+``` nimrod
+proc setName(self: Animal, name: string) =
+  self.name = name
+```
+the error message is a lot clearer:
+``` console
+Error: 'self.name' cannot be assigned to
+```
+
+The reason for this is, that arguments to `proc`s are immutable. This is simply solved by marking mutable arguments using `var` as shown below:
```

Comment:
```suggestion
This happens because arguments to `proc` are immutable by default. The error can be fixed by marking the argument as mutable using `var`:
```